src/key_gen/tests.rs

```diff
@@ -109,50 +110,88 @@ fn all_nodes_being_responsive() -> Result<()> {
-    // With one non_responsive node, Proposal phase cannot be completed automatically. This
-    // requires finalize_contributing_phase to be called externally to complete the procedure.
-    // All participants will transit into Complaint phase afterwards, Then requires
-    // finalize_complaining_phase to be called externally to complete the procedure.
-    for _ in 0..2 {
-        peer_ids.iter().enumerate().for_each(|(index, _peer_id)| {
-            if let Ok(proposal_vec) = generators[index].timed_phase_transition(&mut rng) {
-                if !non_responsives.contains(&(index as u64)) {
-                    for proposal in proposal_vec {
-                        proposals.push(proposal);
+    let combinations_of_non_resp = all_nodes
+        .iter()
+        .cloned()
+        .combinations(NODENUM - THRESHOLD - 1);
+
+    for non_responsive in combinations_of_non_resp {
+        let non_responsive: BTreeSet<u64> = non_responsive.iter().cloned().collect();
+        let (peer_ids, mut generators) = setup_generators(&mut rng, non_responsive.clone())?;
+
+        let mut proposals = Vec::new();
+        // With one non_responsive node, Proposal phase cannot be completed automatically. This
+        // requires finalize_contributing_phase to be called externally to complete the procedure.
+        // All participants will transit into Complaint phase afterwards, Then requires
+        // finalize_complaining_phase to be called externally to complete the procedure.
+        for _ in 0..2 {
+            peer_ids.iter().enumerate().for_each(|(index, _peer_id)| {
+                if let Ok(proposal_vec) = generators[index].timed_phase_transition(&mut rng) {
+                    if !non_responsive.contains(&(index as u64)) {
+                        for proposal in proposal_vec {
+                            proposals.push(proposal);
+                        }
                     }
                 }
-            }
-        });
-        // Continue the procedure with messaging.
-        messaging(
-            &mut rng,
-            &mut generators,
-            &mut proposals,
-            non_responsives.clone(),
-        );
-        assert!(proposals.is_empty());
-    }
+            });
+            // Continue the procedure with messaging.
+            messaging(
+                &mut rng,
+                &mut generators,
+                &mut proposals,
+                non_responsive.clone(),
+            );
+            assert!(proposals.is_empty());
+        }
 
-    generators
-        .iter_mut()
-        .enumerate()
-        .for_each(|(index, key_gen)| {
-            if !non_responsives.contains(&(index as u64)) {
-                assert!(key_gen.generate_keys().is_some());
-                non_responsives.iter().for_each(|idx| {
+        let responsive = all_nodes.difference(&non_responsive).cloned().collect_vec();
+
+        let pub_key_set: PublicKeySet = generators[responsive.as_slice()[0] as usize]
```

Comment:
`as_slice` is not needed. You can index a `Vec` directly: `generators[responsive[0] as usize]`